app/pages/project/index.spec.js

```diff
@@ -49,45 +49,100 @@ const resources = {
   ]
 };
 
-const project = apiClient.type('projects').create({
-  id: 'a',
-  display_name: 'A test project',
-  experimental_tools: [],
-  get(type) {
-    return Promise.resolve(resources[type]);
-  },
-  links: {
-    active_workflows: ['1', '2', '3', '4', '5'],
-    avatar: { id: '1' },
-    background: { id: '1' },
-    owner: { id: '1' }
+function mockPanoptesResource(type, options) {
+  const resource = apiClient.type(type).create(options);
+  if (!resource.save.restore) {
+    sinon.stub(resource, 'save', () => Promise.resolve(resource));
+    sinon.stub(resource, 'get', resourceType => Promise.resolve(resources[resourceType]));
+    sinon.stub(resource, 'delete');
   }
-});
-
-const background = apiClient.type('backgrounds').create({
-  id: project.links.background.id
-});
+  return resource;
+}
+
+const project = mockPanoptesResource('projects',
+  {
+    id: 'a',
+    display_name: 'A test project',
+    experimental_tools: [],
+    links: {
+      active_workflows: ['1', '2', '3', '4', '5'],
+      avatar: { id: '1' },
+      background: { id: '1' },
+      owner: { id: '1' }
+    }
+  }
+);
 
-const owner = apiClient.type('users').create({
-  id: project.links.owner.id
-});
+const background = mockPanoptesResource(
+  'backgrounds',
+  {
+    id: project.links.background.id
+  }
+);
 
-const projectAvatar = apiClient.type('avatars').create({
-  id: project.links.avatar.id
-});
+const owner = mockPanoptesResource(
+  'users',
+  {
+    id: project.links.owner.id
+  }
+);
 
-const preferences = apiClient.type('project_preferences').create({
-  preferences: {},
-  links: {
-    project: project.id
+const projectAvatar = mockPanoptesResource(
+  'avatars',
+  {
+    id: project.links.avatar.id
   }
-});
+);
+
+const preferences = mockPanoptesResource(
+  'project_preferences',
+  {
+    preferences: {},
+    links: {
+      project: project.id
+    }
+  }
+);
 
 describe('ProjectPageController', () => {
-  const wrapper = shallow(<ProjectPageController params={params} project={project} location={location} />, { context });
+  const actions = {
+    translations: {
+      load: () => null
+    }
+  };
+  const translations = {
+    locale: 'en'
+  };
+  const wrapper = shallow(
+    <ProjectPageController
+      actions={actions}
+      params={params}
+      project={project}
+      location={location}
+      translations={translations}
+    />,
+    { context }
+  );
   const controller = wrapper.instance();
   const workflowSpy = sinon.spy(controller, 'getWorkflow');
 
+  before(() => {
+    sinon.stub(apiClient, 'request', (method, url, payload) => {
+      if (url === '/workflows') {
```

Comment:
This branching when setting up the stub looks like the test context could be located closer to the relevant test instead of at the higher level. Having test setup context like this normally signifies this to me and I appreciate it might be hard to setup the context at a finer level.

Reply:
This was a workaround for workflow selection going into an infinite loop if you don't return anything from a network request. I think it affects all the tests, but it's not a very good fix.